# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: gfx/thebes/gfxPlatformFontList.cpp
# Commit: a8b9260fbdef
# Full Hash: a8b9260fbdef0f3600dc0b8e2e245cb0f82c3470
# Author: Jonathan Kew <jkew@mozilla.com>
# Date: 2021-09-08 21:39:05
# Regressor Bug: 1715501
# File Overlap Count: 1
# Description:
#   Bug 1715501 - patch 2 - Pass the presContext to platform font lookup methods so they can query it for font visibility. r=emilio
#   
#   This does not in itself change user-visible behavior, but is a foundation for bug
#   1715507 where we will make the behavior per-context instead of global. This is basically
#   just plumbing, passing a pointer to the presContext that's asking for fonts to be
# ==============================================================================

diff -r bed82cc6a625 -r a8b9260fbdef gfx/thebes/gfxPlatformFontList.cpp
--- a/gfx/thebes/gfxPlatformFontList.cpp	Wed Sep 08 12:18:17 2021 +0000
+++ b/gfx/thebes/gfxPlatformFontList.cpp	Wed Sep 08 12:18:17 2021 +0000
@@ -17,6 +17,7 @@
 #include "GeckoProfiler.h"
 #include "nsCRT.h"
 #include "nsGkAtoms.h"
+#include "nsPresContext.h"
 #include "nsServiceManagerUtils.h"
 #include "nsUnicharUtils.h"
 #include "nsUnicodeProperties.h"
@@ -132,11 +133,8 @@
     {0, 0, 0, {0, 0, 0}}  // terminator
 };
 
-static const char* kObservedPrefs[] = {"font.",
-                                       "font.name-list.",
+static const char* kObservedPrefs[] = {"font.", "font.name-list.",
                                        "intl.accept_languages",  // hmmmm...
-                                       "layout.css.font-visibility.level",
-                                       "privacy.resistFingerprinting",
                                        nullptr};
 
 static const char kFontSystemWhitelistPref[] = "font.system.whitelist";
@@ -166,13 +164,6 @@
   // changed but it probably isn't that big a deal.
   gfxPlatformFontList::PlatformFontList()->ClearLangGroupPrefFonts();
   gfxFontCache::GetCache()->AgeAllGenerations();
-  if (aPref && (!strcmp(aPref, "layout.css.font-visibility.level") ||
-                !strcmp(aPref, "privacy.resistFingerprinting"))) {
-    gfxPlatformFontList::PlatformFontList()->SetVisibilityLevel();
-    if (XRE_IsParentProcess()) {
-      gfxPlatform::ForceGlobalReflow(gfxPlatform::NeedsReframe::No);
-    }
-  }
 }
 
 static gfxFontListPrefObserver* gFontListPrefObserver = nullptr;
@@ -521,8 +512,6 @@
     gfxFontUtils::GetPrefsFontList(kFontSystemWhitelistPref, mEnabledFontsList);
   }
 
-  SetVisibilityLevel();
-
   // From here, gfxPlatformFontList::IsInitialized will return true,
   // unless InitFontListForPlatform() fails and we reset it below.
   mFontlistInitCount++;
@@ -578,7 +567,7 @@
   // Set up mDefaultFontEntry as a "last resort" default that we can use
   // to avoid crashing if the font list is otherwise unusable.
   gfxFontStyle defStyle;
-  FontFamily fam = GetDefaultFont(&defStyle);
+  FontFamily fam = GetDefaultFont(nullptr, &defStyle);
   if (fam.mIsShared) {
     auto face = fam.mShared->FindFaceForStyle(SharedFontList(), defStyle);
     mDefaultFontEntry =
@@ -604,19 +593,6 @@
   }
 }
 
-void gfxPlatformFontList::SetVisibilityLevel() {
-  FontVisibility newLevel;
-  if (StaticPrefs::privacy_resistFingerprinting()) {
-    newLevel = FontVisibility::Base;
-  } else {
-    newLevel = FontVisibility(
-        std::min(int32_t(FontVisibility::User),
-                 std::max(int32_t(FontVisibility::Base),
-                          StaticPrefs::layout_css_font_visibility_level())));
-  }
-  mVisibilityLevel = newLevel;
-}
-
 void gfxPlatformFontList::FontListChanged() {
   MOZ_ASSERT(!XRE_IsParentProcess());
   InitializeCodepointsWithNoFonts();
@@ -789,8 +765,9 @@
 }
 
 gfxFontEntry* gfxPlatformFontList::LookupInSharedFaceNameList(
-    const nsACString& aFaceName, WeightRange aWeightForEntry,
-    StretchRange aStretchForEntry, SlantStyleRange aStyleForEntry) {
+    nsPresContext* aPresContext, const nsACString& aFaceName,
+    WeightRange aWeightForEntry, StretchRange aStretchForEntry,
+    SlantStyleRange aStyleForEntry) {
   nsAutoCString keyName(aFaceName);
   ToLowerCase(keyName);
   fontlist::FontList* list = SharedFontList();
@@ -812,6 +789,11 @@
   if (!face || !family) {
     return nullptr;
   }
+  FontVisibility level =
+      aPresContext ? aPresContext->GetFontVisibility() : FontVisibility::User;
+  if (!IsVisibleToCSS(*family, level)) {
+    return nullptr;
+  }
   gfxFontEntry* fe = CreateFontEntry(face, family);
   if (fe) {
     fe->mIsLocalUserFont = true;
@@ -851,15 +833,14 @@
   }
 }
 
-bool gfxPlatformFontList::IsVisibleToCSS(const gfxFontFamily& aFamily) const {
-  return aFamily.Visibility() <= mVisibilityLevel ||
-         IsFontFamilyWhitelistActive();
+bool gfxPlatformFontList::IsVisibleToCSS(const gfxFontFamily& aFamily,
+                                         FontVisibility aVisibility) const {
+  return aFamily.Visibility() <= aVisibility || IsFontFamilyWhitelistActive();
 }
 
-bool gfxPlatformFontList::IsVisibleToCSS(
-    const fontlist::Family& aFamily) const {
-  return aFamily.Visibility() <= mVisibilityLevel ||
-         IsFontFamilyWhitelistActive();
+bool gfxPlatformFontList::IsVisibleToCSS(const fontlist::Family& aFamily,
+                                         FontVisibility aVisibility) const {
+  return aFamily.Visibility() <= aVisibility || IsFontFamilyWhitelistActive();
 }
 
 void gfxPlatformFontList::GetFontList(nsAtom* aLangGroup,
@@ -871,7 +852,7 @@
     if (families) {
       for (uint32_t i = 0; i < list->NumFamilies(); i++) {
         auto& f = families[i];
-        if (!IsVisibleToCSS(f) || f.IsAltLocaleFamily()) {
+        if (!IsVisibleToCSS(f, FontVisibility::User) || f.IsAltLocaleFamily()) {
           continue;
         }
         // XXX TODO: filter families for aGenericFamily, if supported by
@@ -885,7 +866,7 @@
 
   MutexAutoLock lock(mFontFamiliesMutex);
   for (const RefPtr<gfxFontFamily>& family : mFontFamilies.Values()) {
-    if (!IsVisibleToCSS(*family)) {
+    if (!IsVisibleToCSS(*family, FontVisibility::User)) {
       continue;
     }
     if (family->FilterForFontList(aLangGroup, aGenericFamily)) {
@@ -910,10 +891,12 @@
 }
 
 gfxFont* gfxPlatformFontList::SystemFindFontForChar(
-    uint32_t aCh, uint32_t aNextCh, Script aRunScript,
-    eFontPresentation aPresentation, const gfxFontStyle* aStyle,
-    FontVisibility* aVisibility) {
-  MOZ_ASSERT(!mCodepointsWithNoFonts[mVisibilityLevel].test(aCh),
+    nsPresContext* aPresContext, uint32_t aCh, uint32_t aNextCh,
+    Script aRunScript, eFontPresentation aPresentation,
+    const gfxFontStyle* aStyle, FontVisibility* aVisibility) {
+  FontVisibility level =
+      aPresContext ? aPresContext->GetFontVisibility() : FontVisibility::User;
+  MOZ_ASSERT(!mCodepointsWithNoFonts[level].test(aCh),
              "don't call for codepoints already known to be unsupported");
 
   // Try to short-circuit font fallback for U+FFFD, used to represent
@@ -922,7 +905,7 @@
   // etc.
   if (aCh == 0xFFFD) {
     gfxFontEntry* fontEntry = nullptr;
-    auto& fallbackFamily = mReplacementCharFallbackFamily[mVisibilityLevel];
+    auto& fallbackFamily = mReplacementCharFallbackFamily[level];
     if (fallbackFamily.mIsShared && fallbackFamily.mShared) {
       fontlist::Face* face =
           fallbackFamily.mShared->FindFaceForStyle(SharedFontList(), *aStyle);
@@ -947,8 +930,9 @@
   // search commonly available fonts
   bool common = true;
   FontFamily fallbackFamily;
-  gfxFont* candidate = CommonFontFallback(
-      aCh, aNextCh, aRunScript, aPresentation, aStyle, fallbackFamily);
+  gfxFont* candidate =
+      CommonFontFallback(aPresContext, aCh, aNextCh, aRunScript, aPresentation,
+                         aStyle, fallbackFamily);
   gfxFont* font = nullptr;
   if (candidate) {
     if (aPresentation == eFontPresentation::Any) {
@@ -966,8 +950,8 @@
   uint32_t cmapCount = 0;
   if (!font) {
     common = false;
-    font = GlobalFontFallback(aCh, aNextCh, aRunScript, aPresentation, aStyle,
-                              cmapCount, fallbackFamily);
+    font = GlobalFontFallback(aPresContext, aCh, aNextCh, aRunScript,
+                              aPresentation, aStyle, cmapCount, fallbackFamily);
     // If the font we found doesn't match the requested type, and we also found
     // a candidate above, prefer that one.
     if (font && aPresentation != eFontPresentation::Any && candidate) {
@@ -997,13 +981,13 @@
 
   // no match? add to set of non-matching codepoints
   if (!font) {
-    mCodepointsWithNoFonts[mVisibilityLevel].set(aCh);
+    mCodepointsWithNoFonts[level].set(aCh);
   } else {
     *aVisibility = fallbackFamily.mIsShared
                        ? fallbackFamily.mShared->Visibility()
                        : fallbackFamily.mUnshared->Visibility();
     if (aCh == 0xFFFD) {
-      mReplacementCharFallbackFamily[mVisibilityLevel] = fallbackFamily;
+      mReplacementCharFallbackFamily[level] = fallbackFamily;
     }
   }
 
@@ -1027,17 +1011,20 @@
 #define NUM_FALLBACK_FONTS 8
 
 gfxFont* gfxPlatformFontList::CommonFontFallback(
-    uint32_t aCh, uint32_t aNextCh, Script aRunScript,
-    eFontPresentation aPresentation, const gfxFontStyle* aMatchStyle,
-    FontFamily& aMatchedFamily) {
+    nsPresContext* aPresContext, uint32_t aCh, uint32_t aNextCh,
+    Script aRunScript, eFontPresentation aPresentation,
+    const gfxFontStyle* aMatchStyle, FontFamily& aMatchedFamily) {
   AutoTArray<const char*, NUM_FALLBACK_FONTS> defaultFallbacks;
   gfxPlatform::GetPlatform()->GetCommonFallbackFonts(
       aCh, aRunScript, aPresentation, defaultFallbacks);
   GlobalFontMatch data(aCh, aNextCh, *aMatchStyle, aPresentation);
+  FontVisibility level =
+      aPresContext ? aPresContext->GetFontVisibility() : FontVisibility::User;
   if (SharedFontList()) {
     for (const auto name : defaultFallbacks) {
-      fontlist::Family* family = FindSharedFamily(nsDependentCString(name));
-      if (!family || !IsVisibleToCSS(*family)) {
+      fontlist::Family* family =
+          FindSharedFamily(aPresContext, nsDependentCString(name));
+      if (!family || !IsVisibleToCSS(*family, level)) {
         continue;
       }
       // XXX(jfkthame) Should we fire the async cmap-loader here, or let it
@@ -1052,7 +1039,7 @@
     for (const auto name : defaultFallbacks) {
       gfxFontFamily* fallback =
           FindFamilyByCanonicalName(nsDependentCString(name));
-      if (!fallback || !IsVisibleToCSS(*fallback)) {
+      if (!fallback || !IsVisibleToCSS(*fallback, level)) {
         continue;
       }
       fallback->FindFontForChar(&data);
@@ -1066,19 +1053,22 @@
 }
 
 gfxFont* gfxPlatformFontList::GlobalFontFallback(
-    uint32_t aCh, uint32_t aNextCh, Script aRunScript,
-    eFontPresentation aPresentation, const gfxFontStyle* aMatchStyle,
-    uint32_t& aCmapCount, FontFamily& aMatchedFamily) {
+    nsPresContext* aPresContext, uint32_t aCh, uint32_t aNextCh,
+    Script aRunScript, eFontPresentation aPresentation,
+    const gfxFontStyle* aMatchStyle, uint32_t& aCmapCount,
+    FontFamily& aMatchedFamily) {
   bool useCmaps = IsFontFamilyWhitelistActive() ||
                   gfxPlatform::GetPlatform()->UseCmapsDuringSystemFallback();
+  FontVisibility level =
+      aPresContext ? aPresContext->GetFontVisibility() : FontVisibility::User;
   FontVisibility rejectedFallbackVisibility = FontVisibility::Unknown;
   if (!useCmaps) {
     // Allow platform-specific fallback code to try and find a usable font
-    gfxFontEntry* fe = PlatformGlobalFontFallback(aCh, aRunScript, aMatchStyle,
-                                                  aMatchedFamily);
+    gfxFontEntry* fe = PlatformGlobalFontFallback(aPresContext, aCh, aRunScript,
+                                                  aMatchStyle, aMatchedFamily);
     if (fe) {
       if (aMatchedFamily.mIsShared) {
-        if (IsVisibleToCSS(*aMatchedFamily.mShared)) {
+        if (IsVisibleToCSS(*aMatchedFamily.mShared, level)) {
           gfxFont* font = fe->FindOrMakeFont(aMatchStyle);
           if (font) {
             if (aPresentation == eFontPresentation::Any) {
@@ -1095,7 +1085,7 @@
         }
         rejectedFallbackVisibility = aMatchedFamily.mShared->Visibility();
       } else {
-        if (IsVisibleToCSS(*aMatchedFamily.mUnshared)) {
+        if (IsVisibleToCSS(*aMatchedFamily.mUnshared, level)) {
           gfxFont* font = fe->FindOrMakeFont(aMatchStyle);
           if (font) {
             if (aPresentation == eFontPresentation::Any) {
@@ -1120,7 +1110,7 @@
     if (families) {
       for (uint32_t i = 0; i < SharedFontList()->NumFamilies(); i++) {
         fontlist::Family& family = families[i];
-        if (!IsVisibleToCSS(family)) {
+        if (!IsVisibleToCSS(family, level)) {
           continue;
         }
         if (!family.IsFullyInitialized() &&
@@ -1146,7 +1136,7 @@
     // iterate over all font families to find a font that support the
     // character
     for (const RefPtr<gfxFontFamily>& family : mFontFamilies.Values()) {
-      if (!IsVisibleToCSS(*family)) {
+      if (!IsVisibleToCSS(*family, level)) {
         continue;
       }
       // evaluate all fonts in this family for a match
@@ -1383,13 +1373,16 @@
 }
 
 bool gfxPlatformFontList::FindAndAddFamilies(
-    StyleGenericFontFamily aGeneric, const nsACString& aFamily,
-    nsTArray<FamilyAndGeneric>* aOutput, FindFamiliesFlags aFlags,
-    gfxFontStyle* aStyle, nsAtom* aLanguage, gfxFloat aDevToCssSize) {
+    nsPresContext* aPresContext, StyleGenericFontFamily aGeneric,
+    const nsACString& aFamily, nsTArray<FamilyAndGeneric>* aOutput,
+    FindFamiliesFlags aFlags, gfxFontStyle* aStyle, nsAtom* aLanguage,
+    gfxFloat aDevToCssSize) {
   nsAutoCString key;
   GenerateFontListKey(aFamily, key);
 
   bool allowHidden = bool(aFlags & FindFamiliesFlags::eSearchHiddenFamilies);
+  FontVisibility visibilityLevel =
+      aPresContext ? aPresContext->GetFontVisibility() : FontVisibility::User;
   if (SharedFontList()) {
     fontlist::Family* family = SharedFontList()->FindFamily(key);
     // If not found, and other family names have not yet been initialized,
@@ -1434,8 +1427,8 @@
     }
     // Check whether the family we found is actually allowed to be looked up,
     // according to current font-visibility prefs.
-    if (family &&
-        (IsVisibleToCSS(*family) || (allowHidden && family->IsHidden()))) {
+    if (family && (IsVisibleToCSS(*family, visibilityLevel) ||
+                   (allowHidden && family->IsHidden()))) {
       aOutput->AppendElement(FamilyAndGeneric(family, aGeneric));
       return true;
     }
@@ -1447,7 +1440,7 @@
 
   // lookup in canonical (i.e. English) family name list
   gfxFontFamily* familyEntry = mFontFamilies.GetWeak(key);
-  if (familyEntry && !IsVisibleToCSS(*familyEntry) &&
+  if (familyEntry && !IsVisibleToCSS(*familyEntry, visibilityLevel) &&
       !(allowHidden && familyEntry->IsHidden())) {
     return false;
   }
@@ -1456,7 +1449,7 @@
   if (!familyEntry) {
     familyEntry = mOtherFamilyNames.GetWeak(key);
   }
-  if (familyEntry && !IsVisibleToCSS(*familyEntry) &&
+  if (familyEntry && !IsVisibleToCSS(*familyEntry, visibilityLevel) &&
       !(allowHidden && familyEntry->IsHidden())) {
     return false;
   }
@@ -1479,7 +1472,7 @@
       }
       mOtherNamesMissed->Insert(key);
     }
-    if (familyEntry && !IsVisibleToCSS(*familyEntry) &&
+    if (familyEntry && !IsVisibleToCSS(*familyEntry, visibilityLevel) &&
         !(allowHidden && familyEntry->IsHidden())) {
       return false;
     }
@@ -1504,7 +1497,7 @@
     }
     if (index > 0) {
       gfxFontFamily* base =
-          FindUnsharedFamily(Substring(aFamily, 0, index),
+          FindUnsharedFamily(aPresContext, Substring(aFamily, 0, index),
                              FindFamiliesFlags::eNoSearchForLegacyFamilyNames);
       // If we found the "base" family name, and if it has members with
       // legacy names, this will add corresponding font-family entries to
@@ -1512,7 +1505,7 @@
       if (base && base->CheckForLegacyFamilyNames(this)) {
         familyEntry = mOtherFamilyNames.GetWeak(key);
       }
-      if (familyEntry && !IsVisibleToCSS(*familyEntry) &&
+      if (familyEntry && !IsVisibleToCSS(*familyEntry, visibilityLevel) &&
           !(allowHidden && familyEntry->IsHidden())) {
         return false;
       }
@@ -1528,14 +1521,15 @@
 }
 
 fontlist::Family* gfxPlatformFontList::FindSharedFamily(
-    const nsACString& aFamily, FindFamiliesFlags aFlags, gfxFontStyle* aStyle,
-    nsAtom* aLanguage, gfxFloat aDevToCss) {
+    nsPresContext* aPresContext, const nsACString& aFamily,
+    FindFamiliesFlags aFlags, gfxFontStyle* aStyle, nsAtom* aLanguage,
+    gfxFloat aDevToCss) {
   if (!SharedFontList()) {
     return nullptr;
   }
   AutoTArray<FamilyAndGeneric, 1> families;
-  if (!FindAndAddFamilies(StyleGenericFontFamily::None, aFamily, &families,
-                          aFlags, aStyle, aLanguage, aDevToCss) ||
+  if (!FindAndAddFamilies(aPresContext, StyleGenericFontFamily::None, aFamily,
+                          &families, aFlags, aStyle, aLanguage, aDevToCss) ||
       !families[0].mFamily.mIsShared) {
     return nullptr;
   }
@@ -1635,10 +1629,11 @@
 }
 
 gfxFontEntry* gfxPlatformFontList::FindFontForFamily(
-    const nsACString& aFamily, const gfxFontStyle* aStyle) {
+    nsPresContext* aPresContext, const nsACString& aFamily,
+    const gfxFontStyle* aStyle) {
   nsAutoCString key;
   GenerateFontListKey(aFamily, key);
-  FontFamily family = FindFamily(key);
+  FontFamily family = FindFamily(aPresContext, key);
   if (family.IsNull()) {
     return nullptr;
   }
@@ -1697,11 +1692,11 @@
 
 bool gfxPlatformFontList::GetStandardFamilyName(const nsCString& aFontName,
                                                 nsACString& aFamilyName) {
-  FontFamily family = FindFamily(aFontName);
+  FontFamily family = FindFamily(nullptr, aFontName);
   if (family.IsNull()) {
     return false;
   }
-  return GetLocalizedFamilyName(FindFamily(aFontName), aFamilyName);
+  return GetLocalizedFamilyName(family, aFamilyName);
 }
 
 bool gfxPlatformFontList::GetLocalizedFamilyName(const FontFamily& aFamily,
@@ -1730,7 +1725,7 @@
       NameListPref(aGenericFamily, aLangGroup).get(), names);
 
   for (const nsCString& name : names) {
-    FontFamily family = FindFamily(name);
+    FontFamily family = FindFamily(nullptr, name);
     if (!family.IsNull()) {
       return FamilyAndGeneric(family);
     }
@@ -1801,8 +1796,8 @@
 }
 
 void gfxPlatformFontList::ResolveGenericFontNames(
-    StyleGenericFontFamily aGenericType, eFontPrefLang aPrefLang,
-    PrefFontList* aGenericFamilies) {
+    nsPresContext* aPresContext, StyleGenericFontFamily aGenericType,
+    eFontPrefLang aPrefLang, PrefFontList* aGenericFamilies) {
   const char* langGroupStr = GetPrefLangName(aPrefLang);
   const char* generic = GetGenericName(aGenericType);
 
@@ -1827,8 +1822,8 @@
     GetSystemUIFontFamilies(langGroup, genericFamilies);
   }
 
-  GetFontFamiliesFromGenericFamilies(aGenericType, genericFamilies, langGroup,
-                                     aGenericFamilies);
+  GetFontFamiliesFromGenericFamilies(
+      aPresContext, aGenericType, genericFamilies, langGroup, aGenericFamilies);
 
 #if 0  // dump out generic mappings
     printf("%s ===> ", NamePref(generic, langGroupStr).get());
@@ -1843,26 +1838,26 @@
 }
 
 void gfxPlatformFontList::ResolveEmojiFontNames(
-    PrefFontList* aGenericFamilies) {
+    nsPresContext* aPresContext, PrefFontList* aGenericFamilies) {
   // emoji preference has no lang name
   AutoTArray<nsCString, 4> genericFamilies;
 
   nsAutoCString prefFontListName("font.name-list.emoji");
   gfxFontUtils::AppendPrefsFontList(prefFontListName.get(), genericFamilies);
 
-  GetFontFamiliesFromGenericFamilies(StyleGenericFontFamily::MozEmoji,
-                                     genericFamilies, nullptr,
-                                     aGenericFamilies);
+  GetFontFamiliesFromGenericFamilies(
+      aPresContext, StyleGenericFontFamily::MozEmoji, genericFamilies, nullptr,
+      aGenericFamilies);
 }
 
 void gfxPlatformFontList::GetFontFamiliesFromGenericFamilies(
-    StyleGenericFontFamily aGenericType,
+    nsPresContext* aPresContext, StyleGenericFontFamily aGenericType,
     nsTArray<nsCString>& aGenericNameFamilies, nsAtom* aLangGroup,
     PrefFontList* aGenericFamilies) {
   // lookup and add platform fonts uniquely
   for (const nsCString& genericFamily : aGenericNameFamilies) {
     AutoTArray<FamilyAndGeneric, 10> families;
-    FindAndAddFamilies(aGenericType, genericFamily, &families,
+    FindAndAddFamilies(aPresContext, aGenericType, genericFamily, &families,
                        FindFamiliesFlags(0), nullptr, aLangGroup);
     for (const FamilyAndGeneric& f : families) {
       if (!aGenericFamilies->Contains(f.mFamily)) {
@@ -1873,14 +1868,15 @@
 }
 
 gfxPlatformFontList::PrefFontList* gfxPlatformFontList::GetPrefFontsLangGroup(
-    StyleGenericFontFamily aGenericType, eFontPrefLang aPrefLang) {
+    nsPresContext* aPresContext, StyleGenericFontFamily aGenericType,
+    eFontPrefLang aPrefLang) {
   if (aGenericType == StyleGenericFontFamily::MozEmoji ||
       aPrefLang == eFontPrefLang_Emoji) {
     // Emoji font has no lang
     PrefFontList* prefFonts = mEmojiPrefFont.get();
     if (MOZ_UNLIKELY(!prefFonts)) {
       prefFonts = new PrefFontList;
-      ResolveEmojiFontNames(prefFonts);
+      ResolveEmojiFontNames(aPresContext, prefFonts);
       mEmojiPrefFont.reset(prefFonts);
     }
     return prefFonts;
@@ -1890,15 +1886,15 @@
   PrefFontList* prefFonts = mLangGroupPrefFonts[aPrefLang][index].get();
   if (MOZ_UNLIKELY(!prefFonts)) {
     prefFonts = new PrefFontList;
-    ResolveGenericFontNames(aGenericType, aPrefLang, prefFonts);
+    ResolveGenericFontNames(aPresContext, aGenericType, aPrefLang, prefFonts);
     mLangGroupPrefFonts[aPrefLang][index].reset(prefFonts);
   }
   return prefFonts;
 }
 
 void gfxPlatformFontList::AddGenericFonts(
-    StyleGenericFontFamily aGenericType, nsAtom* aLanguage,
-    nsTArray<FamilyAndGeneric>& aFamilyList) {
+    nsPresContext* aPresContext, StyleGenericFontFamily aGenericType,
+    nsAtom* aLanguage, nsTArray<FamilyAndGeneric>& aFamilyList) {
   // map lang ==> langGroup
   nsAtom* langGroup = GetLangGroup(aLanguage);
 
@@ -1906,7 +1902,8 @@
   eFontPrefLang prefLang = GetFontPrefLangFor(langGroup);
 
   // lookup pref fonts
-  PrefFontList* prefFonts = GetPrefFontsLangGroup(aGenericType, prefLang);
+  PrefFontList* prefFonts =
+      GetPrefFontsLangGroup(aPresContext, aGenericType, prefLang);
 
   if (!prefFonts->IsEmpty()) {
     aFamilyList.SetCapacity(aFamilyList.Length() + prefFonts->Length());
@@ -2289,8 +2286,9 @@
   return StyleGenericFontFamily::Serif;
 }
 
-FontFamily gfxPlatformFontList::GetDefaultFont(const gfxFontStyle* aStyle) {
-  FontFamily family = GetDefaultFontForPlatform(aStyle);
+FontFamily gfxPlatformFontList::GetDefaultFont(nsPresContext* aPresContext,
+                                               const gfxFontStyle* aStyle) {
+  FontFamily family = GetDefaultFontForPlatform(aPresContext, aStyle);
   if (!family.IsNull()) {
     return family;
   }
@@ -2463,8 +2461,9 @@
         mOtherNamesMissed->cbegin(), mOtherNamesMissed->cend(),
         [&](const auto& key) {
           return FindUnsharedFamily(
-              key, (FindFamiliesFlags::eForceOtherFamilyNamesLoading |
-                    FindFamiliesFlags::eNoAddToNamesMissedWhenSearching));
+              nullptr, key,
+              (FindFamiliesFlags::eForceOtherFamilyNamesLoading |
+               FindFamiliesFlags::eNoAddToNamesMissedWhenSearching));
         });
     if (forceReflow) {
       gfxPlatform::ForceGlobalReflow(gfxPlatform::NeedsReframe::No);